internal/common/context.bzl

```diff
@@ -0,0 +1,309 @@
+        allow_files = True,
+        single_file = True,
+        default = Label("//internal/common/actions/create_source_dir:create_source_dir.js"),
+    ),
+    "_internal_nodejs": attr.label(
+        allow_files = True,
+        single_file = True,
+        default = Label("@nodejs//:node"),
+    ),
+    "_internal_packages": attr.label(
+        default = Label("//internal:packages"),
+    ),
+    "_empty_npm_packages": attr.label(
+        allow_files = True,
+        single_file = True,
+        default = Label("//internal/npm_packages/empty:packages"),
+    ),
+}
+
+# Attributes that should be included for any rule that wants to create a JsLibrary
+JS_LIBRARY_ATTRIBUTES = dict(JS_CONTEXT_ATTRIBUTES, **{
+    "srcs": attr.label_list(
+        allow_files = True,
+        mandatory = True,
+    ),
+    "deps": attr.label_list(
+        default = [],
+    ),
+    "module_name": attr.string(),
+})
+
+RULES_NODEJS_MODULE_ATTRIBUTES = {
+    # The official bazel rules use module_name and module_root for non-relative
+    # module mapping. If the module_root is supplied and the module_name is not
+    # present then the module_name is assumed to be the target name. See:
+    # https://github.com/bazelbuild/rules_nodejs/blob/master/internal/common/module_mappings.bzl
+    "module_name": attr.string(),
+    "module_root": attr.string(),
+    # The official bazel rules look for this tag
+    "tags": ["NODE_MODULE_MARKER"],
+}
+
+###############################################################################
+# Helpers
+
+def _js_library_info(js, attr = None):
+    """Create a JsLibrary provider with attr.srcs and the sources from attr.deps
+
+    Args:
+    js: JsContext object
+    attr: rule attributes to extract srcs and deps from
+    """
+
+    if not attr:
+        attr = js.attr
+
+    # The srcs should contain what has been explicitly added for a rule
+    src_files = js._ctx.files.srcs
+
+    # The deps is list of labels that should have providers that we can get sources from
+    deps_attr = getattr(attr, "deps", [])
+
+    transitive_sources = []
+    node_modules_dirs = []
+    dep_modules = []
+    transitive_dep_modules = []
+    dep_module_targets = []
+    transitive_dep_module_targets = []
+
+    # Iterate through the deps to add them to their correct JsLibrary attributes
+    for dep in deps_attr:
+        if JsLibraryInfo in dep:
+            dep_js_library = dep[JsLibraryInfo]
+
+            # The dependency is another JsLibrary
+            transitive_sources.append(dep_js_library.all_sources)
+            transitive_dep_modules.append(dep_js_library.all_dep_modules)
+        elif JsModuleInfo in dep:
+            dep_js_module = dep[JsModuleInfo]
+            transitive_dep_modules.append(dep_js_module.all_dep_modules)
+            transitive_dep_module_targets.append(dep_js_module.all_dep_module_targets)
+        elif hasattr(dep, "tags") and "NODE_MODULES_MARKER" in getattr(dep, "tags"):
```

Comment:
FYI the tag is defined in the other file as `NODE_MODULE_MARKER`, there's an extra `S` here.

Reply:
fixed.